{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DisGeNET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Disease_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Score_gda</th>\n",
       "      <th>Association_Type</th>\n",
       "      <th>Type</th>\n",
       "      <th>Original_DB</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PMID_Year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>7137</td>\n",
       "      <td>0.66</td>\n",
       "      <td>Biomarker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENOMICS_ENGLAND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>EPG5</td>\n",
       "      <td>57724</td>\n",
       "      <td>0.40</td>\n",
       "      <td>Biomarker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENOMICS_ENGLAND</td>\n",
       "      <td>Role of Epg5 in selective neurodegeneration an...</td>\n",
       "      <td>23674064.0</td>\n",
       "      <td>2013.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>FASLG</td>\n",
       "      <td>356</td>\n",
       "      <td>0.30</td>\n",
       "      <td>Biomarker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTD_human</td>\n",
       "      <td>Myocardial Fas ligand expression increases sus...</td>\n",
       "      <td>17943461.0</td>\n",
       "      <td>2008.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>LAMP2</td>\n",
       "      <td>3920</td>\n",
       "      <td>0.41</td>\n",
       "      <td>Biomarker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENOMICS_ENGLAND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>LMNA</td>\n",
       "      <td>4000</td>\n",
       "      <td>0.70</td>\n",
       "      <td>Biomarker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTD_human</td>\n",
       "      <td>Missense mutations in the rod domain of the &lt;s...</td>\n",
       "      <td>10580070.0</td>\n",
       "      <td>1999.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Disease Disease_id   Gene  Gene_id  Score_gda  \\\n",
       "0  Cardiomyopathy, Dilated   C0007193  TNNI3     7137       0.66   \n",
       "1  Cardiomyopathy, Dilated   C0007193   EPG5    57724       0.40   \n",
       "2  Cardiomyopathy, Dilated   C0007193  FASLG      356       0.30   \n",
       "3  Cardiomyopathy, Dilated   C0007193  LAMP2     3920       0.41   \n",
       "4  Cardiomyopathy, Dilated   C0007193   LMNA     4000       0.70   \n",
       "\n",
       "  Association_Type   Type       Original_DB  \\\n",
       "0        Biomarker    NaN  GENOMICS_ENGLAND   \n",
       "1        Biomarker    NaN  GENOMICS_ENGLAND   \n",
       "2        Biomarker    NaN         CTD_human   \n",
       "3        Biomarker    NaN  GENOMICS_ENGLAND   \n",
       "4        Biomarker    NaN         CTD_human   \n",
       "\n",
       "                                            Sentence        PMID  PMID_Year  \n",
       "0                                                NaN         NaN        NaN  \n",
       "1  Role of Epg5 in selective neurodegeneration an...  23674064.0     2013.0  \n",
       "2  Myocardial Fas ligand expression increases sus...  17943461.0     2008.0  \n",
       "3                                                NaN         NaN        NaN  \n",
       "4  Missense mutations in the rod domain of the <s...  10580070.0     1999.0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load Evidence Gene-Disease Associations df\n",
    "evid_df = pd.read_csv('C0007193_disease_gda_evidences_CURATED.tsv', sep='\\t')\n",
    "evid_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Disease_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Gene_Full_Name</th>\n",
       "      <th>Protein_Class</th>\n",
       "      <th>N_diseases_g</th>\n",
       "      <th>DSI_g</th>\n",
       "      <th>DPI_g</th>\n",
       "      <th>pLI</th>\n",
       "      <th>Score_gda</th>\n",
       "      <th>EL_gda</th>\n",
       "      <th>EI_gda</th>\n",
       "      <th>N_PMIDs</th>\n",
       "      <th>N_SNPs_gda</th>\n",
       "      <th>First_Ref</th>\n",
       "      <th>Last_Ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>DMD</td>\n",
       "      <td>1756</td>\n",
       "      <td>P11532</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48</td>\n",
       "      <td>0.477</td>\n",
       "      <td>0.724</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.90</td>\n",
       "      <td>strong</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>2016.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>TNNT2</td>\n",
       "      <td>7139</td>\n",
       "      <td>P45379</td>\n",
       "      <td>troponin T2, cardiac type</td>\n",
       "      <td>cytoskeletal protein</td>\n",
       "      <td>46</td>\n",
       "      <td>0.611</td>\n",
       "      <td>0.241</td>\n",
       "      <td>2.605000e-03</td>\n",
       "      <td>0.90</td>\n",
       "      <td>strong</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>2018.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>TTN</td>\n",
       "      <td>7273</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>titin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>0.534</td>\n",
       "      <td>0.655</td>\n",
       "      <td>6.433300e-84</td>\n",
       "      <td>0.90</td>\n",
       "      <td>strong</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3</td>\n",
       "      <td>201</td>\n",
       "      <td>1995.0</td>\n",
       "      <td>2017.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>ACTC1</td>\n",
       "      <td>70</td>\n",
       "      <td>P68032</td>\n",
       "      <td>actin, alpha, cardiac muscle 1</td>\n",
       "      <td>cytoskeletal protein</td>\n",
       "      <td>21</td>\n",
       "      <td>0.656</td>\n",
       "      <td>0.414</td>\n",
       "      <td>7.760600e-01</td>\n",
       "      <td>0.84</td>\n",
       "      <td>strong</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2002.0</td>\n",
       "      <td>2014.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>MYH6</td>\n",
       "      <td>4624</td>\n",
       "      <td>P13533</td>\n",
       "      <td>myosin heavy chain 6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>0.621</td>\n",
       "      <td>0.552</td>\n",
       "      <td>3.241600e-31</td>\n",
       "      <td>0.84</td>\n",
       "      <td>strong</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>2016.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Disease Disease_id   Gene  Gene_id UniProt  \\\n",
       "0  Cardiomyopathy, Dilated   C0007193    DMD     1756  P11532   \n",
       "1  Cardiomyopathy, Dilated   C0007193  TNNT2     7139  P45379   \n",
       "2  Cardiomyopathy, Dilated   C0007193    TTN     7273  Q8WZ42   \n",
       "3  Cardiomyopathy, Dilated   C0007193  ACTC1       70  P68032   \n",
       "4  Cardiomyopathy, Dilated   C0007193   MYH6     4624  P13533   \n",
       "\n",
       "                   Gene_Full_Name         Protein_Class  N_diseases_g  DSI_g  \\\n",
       "0                      dystrophin                   NaN            48  0.477   \n",
       "1       troponin T2, cardiac type  cytoskeletal protein            46  0.611   \n",
       "2                           titin                   NaN            37  0.534   \n",
       "3  actin, alpha, cardiac muscle 1  cytoskeletal protein            21  0.656   \n",
       "4            myosin heavy chain 6                   NaN            20  0.621   \n",
       "\n",
       "   DPI_g           pLI  Score_gda  EL_gda  EI_gda  N_PMIDs  N_SNPs_gda  \\\n",
       "0  0.724  1.000000e+00       0.90  strong  0.9375        1           0   \n",
       "1  0.241  2.605000e-03       0.90  strong  1.0000        2          11   \n",
       "2  0.655  6.433300e-84       0.90  strong  1.0000        3         201   \n",
       "3  0.414  7.760600e-01       0.84  strong  0.7500        0           1   \n",
       "4  0.552  3.241600e-31       0.84  strong  1.0000        1           0   \n",
       "\n",
       "   First_Ref  Last_Ref  \n",
       "0     1993.0    2016.0  \n",
       "1     2000.0    2018.0  \n",
       "2     1995.0    2017.0  \n",
       "3     2002.0    2014.0  \n",
       "4     2006.0    2016.0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load Summary Gene-Disease Associations df\n",
    "\n",
    "summary_df = pd.read_csv('C0007193_disease_gda_summary_CURATED.tsv', sep='\\t')\n",
    "summary_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DMD\n",
      "TNNT2\n",
      "TTN\n",
      "ACTC1\n",
      "MYH6\n",
      "SGCD\n",
      "DES\n",
      "MYH7\n",
      "CSRP3\n",
      "SCN5A\n",
      "LMNA\n",
      "BAG3\n",
      "TNNI3\n",
      "LDB3\n",
      "PSEN1\n",
      "TMPO\n",
      "PSEN2\n",
      "SDHA\n",
      "ADRB1\n",
      "ADRB2\n",
      "SOD2\n",
      "EGFR\n",
      "ILK\n",
      "NKX2-5\n",
      "FAS\n",
      "PLN\n",
      "TPM1\n",
      "TNNC1\n",
      "MYBPC3\n",
      "TAZ\n",
      "VCL\n",
      "DNAJC19\n",
      "EYA4\n",
      "DSP\n",
      "DOLK\n",
      "FKTN\n",
      "ABCC9\n",
      "DSG2\n",
      "CRYAB\n",
      "ANKRD1\n",
      "RBM20\n",
      "MYPN\n",
      "FHL2\n",
      "RAF1\n",
      "LAMA4\n",
      "LAMP2\n",
      "NEXN\n",
      "NEBL\n",
      "TXNRD2\n",
      "ACTN2\n",
      "GATAD1\n",
      "TCAP\n",
      "PRDM16\n",
      "JUP\n",
      "EPG5\n",
      "HAMP\n",
      "XK\n",
      "ALMS1\n",
      "GPX1\n",
      "RYR2\n",
      "NPPA\n",
      "EMD\n",
      "SLC22A5\n",
      "IDH2\n",
      "PKP2\n",
      "TMEM43\n",
      "NPPB\n",
      "HFE\n",
      "SYNE1\n",
      "CD36\n",
      "WDR12\n",
      "NR3C2\n",
      "CTF1\n",
      "CAVIN4\n",
      "FLNC\n",
      "MYL3\n",
      "MYL2\n",
      "GLA\n",
      "CSRNP1\n",
      "MPO\n",
      "CSF3\n",
      "ABRA\n",
      "SIK1\n",
      "ITGB1\n",
      "PDLIM3\n",
      "PRKAG2\n",
      "SLC40A1\n",
      "DMPK\n",
      "SCN1B\n",
      "TTR\n",
      "UCP1\n",
      "DSC2\n",
      "RENBP\n",
      "FASN\n",
      "ATM\n",
      "FASLG\n",
      "RAC1\n",
      "FHL1\n",
      "SYNE2\n",
      "RAB3GAP2\n",
      "TFR2\n"
     ]
    }
   ],
   "source": [
    "# print genes in order to give them in input to HGNC\n",
    "for gene in list(summary_df.Gene):\n",
    "    print(gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NB\n",
    "\n",
    "Difference btw summary and evidence:\n",
    "\n",
    "* Summary has for each row a different gene.\n",
    "\n",
    "* Evidence has the same genes of summary, but someone is repeated.\n",
    "\n",
    "If we look at the the lenghts of the 2 sets of genes, they are the same!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "101\n",
      "101\n"
     ]
    }
   ],
   "source": [
    "print(len(set(list(summary_df.Gene))))\n",
    "print(len(set(list(evid_df.Gene))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean Summary Dataframe\n",
    "\n",
    "The TMPO gene has an irregular UniProt value. Infact it has two entries separated by ';'. For this reason we change this row, mantaining just one of the two entry. Thenwe add a new row for the same gene but with the other value of entry. At the end we will have for the same gene, 2 rows, which differ only by the UniProt value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Disease_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Gene_Full_Name</th>\n",
       "      <th>Protein_Class</th>\n",
       "      <th>N_diseases_g</th>\n",
       "      <th>DSI_g</th>\n",
       "      <th>DPI_g</th>\n",
       "      <th>pLI</th>\n",
       "      <th>Score_gda</th>\n",
       "      <th>EL_gda</th>\n",
       "      <th>EI_gda</th>\n",
       "      <th>N_PMIDs</th>\n",
       "      <th>N_SNPs_gda</th>\n",
       "      <th>First_Ref</th>\n",
       "      <th>Last_Ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>TMPO</td>\n",
       "      <td>7112</td>\n",
       "      <td>P42166;P42167</td>\n",
       "      <td>thymopoietin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>0.752</td>\n",
       "      <td>0.241</td>\n",
       "      <td>1.773700e-07</td>\n",
       "      <td>0.61</td>\n",
       "      <td>limited</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>2006.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Disease Disease_id  Gene  Gene_id        UniProt  \\\n",
       "15  Cardiomyopathy, Dilated   C0007193  TMPO     7112  P42166;P42167   \n",
       "\n",
       "   Gene_Full_Name Protein_Class  N_diseases_g  DSI_g  DPI_g           pLI  \\\n",
       "15   thymopoietin           NaN             6  0.752  0.241  1.773700e-07   \n",
       "\n",
       "    Score_gda   EL_gda  EI_gda  N_PMIDs  N_SNPs_gda  First_Ref  Last_Ref  \n",
       "15       0.61  limited     1.0        1           0     2006.0    2006.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# looking at TMPO gene row\n",
    "summary_df[summary_df['Gene'] == 'TMPO']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of UniProt col values\n",
    "col_uniprot = list(summary_df.UniProt)\n",
    "# value corresponfing to TMPO gene\n",
    "double_entries_list = col_uniprot[15].split(';')\n",
    "# first UniProt entry\n",
    "first_entry = double_entries_list[0]\n",
    "# Second UniProt entry\n",
    "second_entry = double_entries_list[1]\n",
    "# updating UniProt column with just the first entry for TMPO gene \n",
    "col_uniprot[15] = first_entry\n",
    "summary_df.UniProt = col_uniprot\n",
    "\n",
    "# listing the TMPO row\n",
    "row_to_update = list((summary_df.loc[15]))\n",
    "# updating the UniProt value with second entry\n",
    "row_to_update[4] = second_entry\n",
    "# adding the new row to df\n",
    "summary_df.loc[len(summary_df)] = row_to_update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Disease_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Gene_Full_Name</th>\n",
       "      <th>Protein_Class</th>\n",
       "      <th>N_diseases_g</th>\n",
       "      <th>DSI_g</th>\n",
       "      <th>DPI_g</th>\n",
       "      <th>pLI</th>\n",
       "      <th>Score_gda</th>\n",
       "      <th>EL_gda</th>\n",
       "      <th>EI_gda</th>\n",
       "      <th>N_PMIDs</th>\n",
       "      <th>N_SNPs_gda</th>\n",
       "      <th>First_Ref</th>\n",
       "      <th>Last_Ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>TMPO</td>\n",
       "      <td>7112</td>\n",
       "      <td>P42166</td>\n",
       "      <td>thymopoietin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>0.752</td>\n",
       "      <td>0.241</td>\n",
       "      <td>1.773700e-07</td>\n",
       "      <td>0.61</td>\n",
       "      <td>limited</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>2006.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>101</td>\n",
       "      <td>Cardiomyopathy, Dilated</td>\n",
       "      <td>C0007193</td>\n",
       "      <td>TMPO</td>\n",
       "      <td>7112</td>\n",
       "      <td>P42167</td>\n",
       "      <td>thymopoietin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>0.752</td>\n",
       "      <td>0.241</td>\n",
       "      <td>1.773700e-07</td>\n",
       "      <td>0.61</td>\n",
       "      <td>limited</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>2006.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Disease Disease_id  Gene  Gene_id UniProt Gene_Full_Name  \\\n",
       "15   Cardiomyopathy, Dilated   C0007193  TMPO     7112  P42166   thymopoietin   \n",
       "101  Cardiomyopathy, Dilated   C0007193  TMPO     7112  P42167   thymopoietin   \n",
       "\n",
       "    Protein_Class  N_diseases_g  DSI_g  DPI_g           pLI  Score_gda  \\\n",
       "15            NaN             6  0.752  0.241  1.773700e-07       0.61   \n",
       "101           NaN             6  0.752  0.241  1.773700e-07       0.61   \n",
       "\n",
       "      EL_gda  EI_gda  N_PMIDs  N_SNPs_gda  First_Ref  Last_Ref  \n",
       "15   limited     1.0        1           0     2006.0    2006.0  \n",
       "101  limited     1.0        1           0     2006.0    2006.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now TMPO gene has 2 rows\n",
    "summary_df[summary_df['Gene'] == 'TMPO']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HGNC\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HGNC dataset \n",
    "hgnc_df = pd.read_csv('hgnc-symbol-check.csv', header= 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input</th>\n",
       "      <th>Match type</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>DMD</td>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>DMD</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>Xp21.2-p21.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>TNNT2</td>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>TNNT2</td>\n",
       "      <td>troponin T2, cardiac type</td>\n",
       "      <td>HGNC:11949</td>\n",
       "      <td>1q32.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>TTN</td>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>TTN</td>\n",
       "      <td>titin</td>\n",
       "      <td>HGNC:12403</td>\n",
       "      <td>2q31.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ACTC1</td>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>ACTC1</td>\n",
       "      <td>actin alpha cardiac muscle 1</td>\n",
       "      <td>HGNC:143</td>\n",
       "      <td>15q14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>MYH6</td>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>MYH6</td>\n",
       "      <td>myosin heavy chain 6</td>\n",
       "      <td>HGNC:7576</td>\n",
       "      <td>14q11.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Input       Match type Approved symbol                 Approved name  \\\n",
       "0    DMD  Approved symbol             DMD                    dystrophin   \n",
       "1  TNNT2  Approved symbol           TNNT2     troponin T2, cardiac type   \n",
       "2    TTN  Approved symbol             TTN                         titin   \n",
       "3  ACTC1  Approved symbol           ACTC1  actin alpha cardiac muscle 1   \n",
       "4   MYH6  Approved symbol            MYH6          myosin heavy chain 6   \n",
       "\n",
       "      HGNC ID      Location  \n",
       "0   HGNC:2928  Xp21.2-p21.1  \n",
       "1  HGNC:11949        1q32.1  \n",
       "2  HGNC:12403        2q31.2  \n",
       "3    HGNC:143         15q14  \n",
       "4   HGNC:7576       14q11.2  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All genes are approved! They are 101\n"
     ]
    }
   ],
   "source": [
    "# save approved genes\n",
    "approved_genes = []\n",
    "for index,row in hgnc_df.iterrows():\n",
    "    if row['Match type'] == 'Approved symbol':\n",
    "        approved_genes.append(row['Input'])\n",
    "        \n",
    "#check if all my input genes are approved\n",
    "if len(set(approved_genes)) == len(set(summary_df.Gene)):\n",
    "    print('All genes are approved! They are', len(set(approved_genes)))\n",
    "else:\n",
    "    print('Not all genes are approved. Check!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Gene names  (primary )</th>\n",
       "      <th>Cross-reference (GeneID)</th>\n",
       "      <th>Gene ontology (molecular function)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>X5DQZ7</td>\n",
       "      <td>Glutathione peroxidase 1 isoform A</td>\n",
       "      <td>GPX1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>GPX1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>glutathione peroxidase activity [GO:0004602]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>X5DQM5</td>\n",
       "      <td>Adrenoceptor beta 2 surface isoform A (Fragment)</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>adenylate cyclase binding [GO:0008179]; beta2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>X5DNJ6</td>\n",
       "      <td>Dystrophia myotonica-protein kinase isoform A ...</td>\n",
       "      <td>DMPK</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>DMPK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATP binding [GO:0005524]; protein serine/threo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>X5D926</td>\n",
       "      <td>Glutathione peroxidase 1 isoform A</td>\n",
       "      <td>GPX1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>GPX1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>glutathione peroxidase activity [GO:0004602]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>X5D8Z6</td>\n",
       "      <td>Dystrophia myotonica-protein kinase isoform B ...</td>\n",
       "      <td>DMPK</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>DMPK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATP binding [GO:0005524]; protein serine/threo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry                                      Protein names Gene names  \\\n",
       "0  X5DQZ7                 Glutathione peroxidase 1 isoform A       GPX1   \n",
       "1  X5DQM5   Adrenoceptor beta 2 surface isoform A (Fragment)      ADRB2   \n",
       "2  X5DNJ6  Dystrophia myotonica-protein kinase isoform A ...       DMPK   \n",
       "3  X5D926                 Glutathione peroxidase 1 isoform A       GPX1   \n",
       "4  X5D8Z6  Dystrophia myotonica-protein kinase isoform B ...       DMPK   \n",
       "\n",
       "               Organism Gene names  (primary ) Cross-reference (GeneID)  \\\n",
       "0  Homo sapiens (Human)                   GPX1                      NaN   \n",
       "1  Homo sapiens (Human)                  ADRB2                      NaN   \n",
       "2  Homo sapiens (Human)                   DMPK                      NaN   \n",
       "3  Homo sapiens (Human)                   GPX1                      NaN   \n",
       "4  Homo sapiens (Human)                   DMPK                      NaN   \n",
       "\n",
       "                  Gene ontology (molecular function)  \n",
       "0       glutathione peroxidase activity [GO:0004602]  \n",
       "1  adenylate cyclase binding [GO:0008179]; beta2-...  \n",
       "2  ATP binding [GO:0005524]; protein serine/threo...  \n",
       "3       glutathione peroxidase activity [GO:0004602]  \n",
       "4  ATP binding [GO:0005524]; protein serine/threo...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read UniProt dataset\n",
    "uni_df = pd.read_excel('uniprot-yourlist_M20191221E5A08BB0B2D1C45B0C7BC3B55FD2655669E875N+AND+organism--.xlsx')\n",
    "uni_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dict: for each gene, uniprot entry\n",
    "gene_uniprot = {}\n",
    "for index,row in summary_df.iterrows():\n",
    "    key = row['Gene']\n",
    "    if key not in list(gene_uniprot.keys()):\n",
    "        gene_uniprot[key] = [row['UniProt']]\n",
    "    else:\n",
    "        gene_uniprot[key].append(row['UniProt'])\n",
    "        \n",
    "# list of uniprot values\n",
    "uniprot_list = [item for sublist in list(gene_uniprot.values()) for item in sublist]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Gene names  (primary )</th>\n",
       "      <th>Cross-reference (GeneID)</th>\n",
       "      <th>Gene ontology (molecular function)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>Q9UPQ8</td>\n",
       "      <td>Dolichol kinase (EC 2.7.1.108) (Transmembrane ...</td>\n",
       "      <td>DOLK KIAA1094 TMEM15 UNQ2422/PRO4980</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>DOLK</td>\n",
       "      <td>22845;</td>\n",
       "      <td>dolichol kinase activity [GO:0004168]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>23</td>\n",
       "      <td>Q9UP52</td>\n",
       "      <td>Transferrin receptor protein 2 (TfR2)</td>\n",
       "      <td>TFR2</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>TFR2</td>\n",
       "      <td>7036;</td>\n",
       "      <td>co-receptor binding [GO:0039706]; transferrin ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>Q9UGJ0</td>\n",
       "      <td>5'-AMP-activated protein kinase subunit gamma-...</td>\n",
       "      <td>PRKAG2</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>PRKAG2</td>\n",
       "      <td>51422;</td>\n",
       "      <td>ADP binding [GO:0043531]; AMP-activated protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35</td>\n",
       "      <td>Q9NP59</td>\n",
       "      <td>Solute carrier family 40 member 1 (Ferroportin...</td>\n",
       "      <td>SLC40A1 FPN1 IREG1 SLC11A3 MSTP079</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>SLC40A1</td>\n",
       "      <td>30061;</td>\n",
       "      <td>ferrous iron transmembrane transporter activit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36</td>\n",
       "      <td>Q9NNW7</td>\n",
       "      <td>Thioredoxin reductase 2, mitochondrial (EC 1.8...</td>\n",
       "      <td>TXNRD2 KIAA1652 TRXR2</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>TXNRD2</td>\n",
       "      <td>10587;</td>\n",
       "      <td>electron transfer activity [GO:0009055]; flavi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Entry                                      Protein names  \\\n",
       "20  Q9UPQ8  Dolichol kinase (EC 2.7.1.108) (Transmembrane ...   \n",
       "23  Q9UP52              Transferrin receptor protein 2 (TfR2)   \n",
       "31  Q9UGJ0  5'-AMP-activated protein kinase subunit gamma-...   \n",
       "35  Q9NP59  Solute carrier family 40 member 1 (Ferroportin...   \n",
       "36  Q9NNW7  Thioredoxin reductase 2, mitochondrial (EC 1.8...   \n",
       "\n",
       "                              Gene names              Organism  \\\n",
       "20  DOLK KIAA1094 TMEM15 UNQ2422/PRO4980  Homo sapiens (Human)   \n",
       "23                                  TFR2  Homo sapiens (Human)   \n",
       "31                                PRKAG2  Homo sapiens (Human)   \n",
       "35    SLC40A1 FPN1 IREG1 SLC11A3 MSTP079  Homo sapiens (Human)   \n",
       "36                 TXNRD2 KIAA1652 TRXR2  Homo sapiens (Human)   \n",
       "\n",
       "   Gene names  (primary ) Cross-reference (GeneID)  \\\n",
       "20                   DOLK                   22845;   \n",
       "23                   TFR2                    7036;   \n",
       "31                 PRKAG2                   51422;   \n",
       "35                SLC40A1                   30061;   \n",
       "36                 TXNRD2                   10587;   \n",
       "\n",
       "                   Gene ontology (molecular function)  \n",
       "20              dolichol kinase activity [GO:0004168]  \n",
       "23  co-receptor binding [GO:0039706]; transferrin ...  \n",
       "31  ADP binding [GO:0043531]; AMP-activated protei...  \n",
       "35  ferrous iron transmembrane transporter activit...  \n",
       "36  electron transfer activity [GO:0009055]; flavi...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filtering uniprot df by uniprot entries which appear also in summary df\n",
    "final_df=uni_df.loc[uni_df['Entry'].isin(uniprot_list)]\n",
    "final_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "102"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check lenght: ok because one gene has 2 different entries from summary df\n",
    "len(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop Gene names col: we have primal gene names, it's sufficient\n",
    "final_df = final_df.drop(['Gene names'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot AC</th>\n",
       "      <th>Protein_Name</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Entrez_ID</th>\n",
       "      <th>Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Q9UPQ8</td>\n",
       "      <td>Dolichol kinase (EC 2.7.1.108) (Transmembrane ...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>DOLK</td>\n",
       "      <td>22845;</td>\n",
       "      <td>dolichol kinase activity [GO:0004168]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Q9UP52</td>\n",
       "      <td>Transferrin receptor protein 2 (TfR2)</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>TFR2</td>\n",
       "      <td>7036;</td>\n",
       "      <td>co-receptor binding [GO:0039706]; transferrin ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Q9UGJ0</td>\n",
       "      <td>5'-AMP-activated protein kinase subunit gamma-...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>PRKAG2</td>\n",
       "      <td>51422;</td>\n",
       "      <td>ADP binding [GO:0043531]; AMP-activated protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Q9NP59</td>\n",
       "      <td>Solute carrier family 40 member 1 (Ferroportin...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>SLC40A1</td>\n",
       "      <td>30061;</td>\n",
       "      <td>ferrous iron transmembrane transporter activit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Q9NNW7</td>\n",
       "      <td>Thioredoxin reductase 2, mitochondrial (EC 1.8...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>TXNRD2</td>\n",
       "      <td>10587;</td>\n",
       "      <td>electron transfer activity [GO:0009055]; flavi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Uniprot AC                                       Protein_Name  \\\n",
       "0     Q9UPQ8  Dolichol kinase (EC 2.7.1.108) (Transmembrane ...   \n",
       "1     Q9UP52              Transferrin receptor protein 2 (TfR2)   \n",
       "2     Q9UGJ0  5'-AMP-activated protein kinase subunit gamma-...   \n",
       "3     Q9NP59  Solute carrier family 40 member 1 (Ferroportin...   \n",
       "4     Q9NNW7  Thioredoxin reductase 2, mitochondrial (EC 1.8...   \n",
       "\n",
       "               Organism Gene_Symbol Entrez_ID  \\\n",
       "0  Homo sapiens (Human)        DOLK    22845;   \n",
       "1  Homo sapiens (Human)        TFR2     7036;   \n",
       "2  Homo sapiens (Human)      PRKAG2    51422;   \n",
       "3  Homo sapiens (Human)     SLC40A1    30061;   \n",
       "4  Homo sapiens (Human)      TXNRD2    10587;   \n",
       "\n",
       "                                            Function  \n",
       "0              dolichol kinase activity [GO:0004168]  \n",
       "1  co-receptor binding [GO:0039706]; transferrin ...  \n",
       "2  ADP binding [GO:0043531]; AMP-activated protei...  \n",
       "3  ferrous iron transmembrane transporter activit...  \n",
       "4  electron transfer activity [GO:0009055]; flavi...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#renaming cols\n",
    "original_cols = list(final_df.columns)\n",
    "new_cols = ['Uniprot AC', 'Protein_Name', 'Organism', 'Gene_Symbol', 'Entrez_ID', 'Function']\n",
    "final_df = final_df.rename(columns=dict(zip(original_cols, new_cols)))\n",
    "# resetting indeces\n",
    "final_df.reset_index(drop=True, inplace= True)\n",
    "final_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "101"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check lenght without duplicates: ok\n",
    "len(set(final_df.Gene_Symbol))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving to csv\n",
    "try:\n",
    "    os.remove('data.csv')\n",
    "except:\n",
    "    pass\n",
    "\n",
    "final_df.to_csv('data.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt1</th>\n",
       "      <th>UniProt2</th>\n",
       "      <th>symbol1</th>\n",
       "      <th>symbol2</th>\n",
       "      <th>evidence type</th>\n",
       "      <th>heart</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>O15273</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>TCAP</td>\n",
       "      <td>TTN</td>\n",
       "      <td>exp;pred</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>O15273</td>\n",
       "      <td>P50461</td>\n",
       "      <td>TCAP</td>\n",
       "      <td>CSRP3</td>\n",
       "      <td>exp;pred</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>O15273</td>\n",
       "      <td>P13929</td>\n",
       "      <td>TCAP</td>\n",
       "      <td>ENO3</td>\n",
       "      <td>exp;pred</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>O15273</td>\n",
       "      <td>P06733</td>\n",
       "      <td>TCAP</td>\n",
       "      <td>ENO1</td>\n",
       "      <td>exp</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>O15273</td>\n",
       "      <td>Q9NP98</td>\n",
       "      <td>TCAP</td>\n",
       "      <td>MYOZ1</td>\n",
       "      <td>exp;pred</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniProt1 UniProt2 symbol1 symbol2 evidence type  heart\n",
       "0   O15273   Q8WZ42    TCAP     TTN      exp;pred      1\n",
       "1   O15273   P50461    TCAP   CSRP3      exp;pred      1\n",
       "2   O15273   P13929    TCAP    ENO3      exp;pred      1\n",
       "3   O15273   P06733    TCAP    ENO1           exp      1\n",
       "4   O15273   Q9NP98    TCAP   MYOZ1      exp;pred      1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iid_df = pd.read_csv('PPIs.txt', sep='\\t')\n",
    "iid_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case we must use experimental data: if we find the tags: e.g. \"experimental;predicted;ortolog\", they ARE considered as experimental."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changing exp;ortho in exp\n",
    "iid_df['evidence type'] = iid_df['evidence type'].replace('exp;ortho', 'exp')\n",
    "#changing exp;ortho;pred in exp\n",
    "iid_df['evidence type'] = iid_df['evidence type'].replace('exp;ortho;pred', 'exp')\n",
    "#changing exp;pred in exp\n",
    "iid_df['evidence type'] = iid_df['evidence type'].replace('exp;pred', 'exp')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We must consider also interactions among no-seed proteins.\n",
    "Then I must consider the no-seed genes among the symbol 1 and symbol 2 cols and find their interactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have in total  6952  interactions. Among them,  0  are among no-seed proteins\n"
     ]
    }
   ],
   "source": [
    "# list of genes in the first symbols col\n",
    "symbols_1 = list(iid_df.symbol1)\n",
    "# list of genes in the second symbols col\n",
    "symbols_2 = list(iid_df.symbol2)\n",
    "\n",
    "#verify that interactions involve always at least one seed gene\n",
    "count = 0\n",
    "for i in range(len(symbols_1)):\n",
    "    if symbols_1[i] not in approved_genes and symbols_2[i] not in approved_genes:\n",
    "        count+=1\n",
    "    \n",
    "print('We have in total ',len(iid_df), ' interactions. Among them, ', count,' are among no-seed proteins')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We found  93  seed genes\n",
      "We have a total of  3945  proteins\n"
     ]
    }
   ],
   "source": [
    "tot_proteins = symbols_1 + symbols_2 \n",
    "tot_proteins = list(set(tot_proteins))\n",
    "\n",
    "seeds_iid_found = []\n",
    "for gene in tot_proteins:\n",
    "    if gene in approved_genes:\n",
    "        seeds_iid_found.append(gene)\n",
    "    \n",
    "seeds_iid_found = list(set(seeds_iid_found))\n",
    "\n",
    "print('We found ', len(seeds_iid_found), ' seed genes')\n",
    "print('We have a total of ', len(tot_proteins), ' proteins')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BIOGRID\n",
    "\n",
    "Here we download the file of human organism from biogrid, and consider just the rows where appear at least one of our seed-genes. So on this way we should consider interactions among only seed-genes and among seed and no-seed genes.\n",
    "\n",
    "Then we will check if there also interactions among only these no-seed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#BioGRID Interaction ID</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>BioGRID ID Interactor A</th>\n",
       "      <th>BioGRID ID Interactor B</th>\n",
       "      <th>Systematic Name Interactor A</th>\n",
       "      <th>Systematic Name Interactor B</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>...</th>\n",
       "      <th>Pubmed ID</th>\n",
       "      <th>Organism Interactor A</th>\n",
       "      <th>Organism Interactor B</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Score</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Phenotypes</th>\n",
       "      <th>Qualifications</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Source Database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>...</td>\n",
       "      <td>9006895</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>...</td>\n",
       "      <td>11309420</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>...</td>\n",
       "      <td>8599089</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>...</td>\n",
       "      <td>10938104</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>...</td>\n",
       "      <td>10875894</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #BioGRID Interaction ID  Entrez Gene Interactor A  \\\n",
       "0                      103                      6416   \n",
       "1                      117                     84665   \n",
       "2                      183                        90   \n",
       "3                      278                      2624   \n",
       "4                      418                      6118   \n",
       "\n",
       "   Entrez Gene Interactor B  BioGRID ID Interactor A  BioGRID ID Interactor B  \\\n",
       "0                      2318                   112315                   108607   \n",
       "1                        88                   124185                   106603   \n",
       "2                      2339                   106605                   108625   \n",
       "3                      5371                   108894                   111384   \n",
       "4                      6774                   112038                   112651   \n",
       "\n",
       "  Systematic Name Interactor A Systematic Name Interactor B  \\\n",
       "0                            -                            -   \n",
       "1                            -                            -   \n",
       "2                            -                            -   \n",
       "3                            -                            -   \n",
       "4                  RP4-547C9.3                            -   \n",
       "\n",
       "  Official Symbol Interactor A Official Symbol Interactor B  \\\n",
       "0                       MAP2K4                         FLNC   \n",
       "1                         MYPN                        ACTN2   \n",
       "2                        ACVR1                         FNTA   \n",
       "3                        GATA2                          PML   \n",
       "4                         RPA2                        STAT3   \n",
       "\n",
       "                               Synonyms Interactor A  ... Pubmed ID  \\\n",
       "0  JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...  ...   9006895   \n",
       "1                             CMD1DD|CMH22|MYOP|RCM4  ...  11309420   \n",
       "2            ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI  ...   8599089   \n",
       "3                           DCML|IMD21|MONOMAC|NFE1B  ...  10938104   \n",
       "4                      REPA2|RP-A p32|RP-A p34|RPA32  ...  10875894   \n",
       "\n",
       "  Organism Interactor A Organism Interactor B      Throughput  Score  \\\n",
       "0                  9606                  9606  Low Throughput      -   \n",
       "1                  9606                  9606  Low Throughput      -   \n",
       "2                  9606                  9606  Low Throughput      -   \n",
       "3                  9606                  9606  Low Throughput      -   \n",
       "4                  9606                  9606  Low Throughput      -   \n",
       "\n",
       "   Modification  Phenotypes Qualifications Tags Source Database  \n",
       "0             -           -              -    -         BIOGRID  \n",
       "1             -           -              -    -         BIOGRID  \n",
       "2             -           -              -    -         BIOGRID  \n",
       "3             -           -              -    -         BIOGRID  \n",
       "4             -           -              -    -         BIOGRID  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biogrid_df = pd.read_table('BIOGRID-ORGANISM-Homo_sapiens-3.5.179.tab2.txt', low_memory = False)\n",
    "biogrid_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "seeds_interactions_A = biogrid_df[biogrid_df['Official Symbol Interactor A'].isin(approved_genes)]\n",
    "seeds_interactions_B= biogrid_df[biogrid_df['Official Symbol Interactor B'].isin(approved_genes)]\n",
    "seeds_interactions = pd.concat([seeds_interactions_A, seeds_interactions_B], axis = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of proteins (including seed genes):  5034\n"
     ]
    }
   ],
   "source": [
    "A = list(set(list(seeds_interactions['Official Symbol Interactor A'])))\n",
    "B = list(set(list(seeds_interactions['Official Symbol Interactor B'])))\n",
    "tot_A_B = set(A + B)\n",
    "tot_len = len(tot_A_B)\n",
    "\n",
    "print('Total number of proteins (including seed genes): ', tot_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of No-Seed Genes:  4936\n",
      "Number of Seed Genes:  98\n"
     ]
    }
   ],
   "source": [
    "no_seed_genes = []\n",
    "for i, row in seeds_interactions.iterrows():\n",
    "    if row['Official Symbol Interactor A'] not in approved_genes:\n",
    "        no_seed_genes.append(row['Official Symbol Interactor A'])\n",
    "    elif row['Official Symbol Interactor B'] not in approved_genes:\n",
    "        no_seed_genes.append(row['Official Symbol Interactor B'])\n",
    "no_seed_genes = list(set(no_seed_genes))\n",
    "print('Number of No-Seed Genes: ', len(set(no_seed_genes))) \n",
    "n_seeds = tot_len - len(no_seed_genes)\n",
    "print('Number of Seed Genes: ', n_seeds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have in total  12979 interactions. Among them,  0  are among no-seed genes\n"
     ]
    }
   ],
   "source": [
    "c = 0\n",
    "symbols_A = list(seeds_interactions['Official Symbol Interactor A'])\n",
    "symbols_B = list(seeds_interactions['Official Symbol Interactor B'])\n",
    "for i in range(len(seeds_interactions)):\n",
    "    if symbols_A[i] in no_seed_genes and symbols_B[i] in no_seed_genes:\n",
    "        c +=1\n",
    "print(\"We have in total \", len(seeds_interactions), 'interactions. Among them, ', c,' are among no-seed genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to csv\n",
    "biogrid_interactions = biogrid_df[['Official Symbol Interactor A', 'Official Symbol Interactor B']]\n",
    "try:\n",
    "    os.remove('biogrid_interactions.csv')\n",
    "except:\n",
    "    pass\n",
    "\n",
    "biogrid_interactions.to_csv('biogrid_interactions.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 1.2 Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_seeds_iid = "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
